### PR TITLE
Add Multigraph with topological sort and cycle breaking

### PR DESCRIPTION
**Pull request by mikary**
This is part of the fix for #1453. It does not improve the error messages, but with cycle breaking a verbose cycle message may not be necessary
@bricelam @maumar

### CODE REVIEW

**Comment by bricelam** [comment]
Looks good. Want to take a stab at updating `ModelDiffer` to break cycles? [Here](https://github.com/aspnet/EntityFramework/blob/dev/src/EntityFramework.Relational/Migrations/Infrastructure/ModelDiffer.cs#L127-L146) is the current implementation.

File: src/EntityFramework.Core/Utilities/Multigraph.cs [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+
+                    var canidateVertices = predecessorCounts.Keys.ToList();
+                    var canidateIndex = 0;
+
+                    // Iterrate over the unsorted verticies
+                    while (canidateIndex < canidateVertices.Count && !broken && canBreakEdge != null)
+                    {
+                        var canidateVertex = canidateVertices[canidateIndex];
+
+                        // Find verticies in the unsorted portion of the graph that have edges to the canidate
```
Comment (divega) [inline src/EntityFramework.Core/Utilities/Multigraph.cs]: typo "canidate"

**Comment by divega** [comment]
@mikary This PR just adds some building blocks and doesn't attempt to fix any issues, correct?I would like to see this applied to solving the problem before it gets checked in.

**Comment by mikary** [comment]
@bricelam @divega I added basic cycle breaking logic to the ModelDiffer, though it is perhaps overly defensive and I haven't added cycle error messages yet. Is this about what we want the pattern to look like?

File: src/EntityFramework.Relational/Migrations/Infrastructure/ModelDiffer.cs [hunk 2]
```diff
@@ -140,14 +139,27 @@ protected virtual IReadOnlyList<MigrationOperation> Sort([NotNull] IEnumerable<M
...
+                        {
+                            alterOperations.Add(foreignKeyOperation);
+                        }
+                        else
+                        {
+                            return false;
+                        }
+                    }
+                    return true;
+                }).Select(k => createTableOperations[k]))
```
Comment (bricelam) [inline src/EntityFramework.Relational/Migrations/Infrastructure/ModelDiffer.cs]: This expression got complex. You should probably store the sorted create table operations in a temporary variable before the return statement.

File: src/EntityFramework.Relational/Migrations/Infrastructure/ModelDiffer.cs [hunk 3]
```diff
@@ -140,14 +139,27 @@ protected virtual IReadOnlyList<MigrationOperation> Sort([NotNull] IEnumerable<M
...
             }
 
             return dropOperations
                 .Concat(createSequenceOperations)
-                .Concat(createTableGraph.TopologicalSort().SelectMany(l => l).Select(k => createTableOperations[k]))
+                .Concat(createTableGraph.TopologicalSort((principalTable, key, foreignKeyOperations) =>
+                {
+                    foreach (var foreignKeyOperation in foreignKeyOperations)
+                    {
+                        if (createTableOperations[key].ForeignKeys.Remove(foreignKeyOperation))
```
Comment (bricelam) [inline src/EntityFramework.Relational/Migrations/Infrastructure/ModelDiffer.cs]: Is this ever going to return `false`?

**Comment by bricelam** [comment]
You should add (at lease) a simple test showing that the model differ can break foreign key cycles.

File: src/EntityFramework.Relational/Migrations/Infrastructure/ModelDiffer.cs (hunk 3 above)
Comment (mikary) [inline src/EntityFramework.Relational/Migrations/Infrastructure/ModelDiffer.cs]: It shouldn't, the keys are all ones we've already iterated over previously.

File: src/EntityFramework.Core/Utilities/Multigraph.cs [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+                            var incommingNeighbour = incommingNeighbours[neighbourIndex];
+                            // Check to see if the edge can be broken
+                            if (canBreakEdge(incommingNeighbour, candidateVertex, _successorMap[incommingNeighbour][candidateVertex]))
+                            {
+                                predecessorCounts[candidateVertex]--;
+                                if (predecessorCounts[candidateVertex] == 0)
+                                {
+                                    sortedQueue.Add(candidateVertex);
+                                    predecessorCounts.Remove(candidateVertex);
+                                    broken = true;
```
Comment (AndriySvyryd) [inline src/EntityFramework.Core/Utilities/Multigraph.cs]: You can replace the inner while with a foreach and use break statement

**Comment by mikary** [comment]
Updated based on PR feedback and added unit test for ```ModelDiffer``` Also fixed key not found issue #1454 in ```BidirectionalAdjacencyListGraph``` since I was adding comments to that file anyway.

**Comment by AndriySvyryd** [comment]
:shipit:

File: src/EntityFramework.Relational/Migrations/Infrastructure/ModelDiffer.cs [hunk 5]
```diff
@@ -140,14 +139,24 @@ protected virtual IReadOnlyList<MigrationOperation> Sort([NotNull] IEnumerable<M
...
+                (principalTable, key, foreignKeyOperations) =>
+                {
+                    foreach (var foreignKeyOperation in foreignKeyOperations)
+                    {
+                        createTableOperations[key].ForeignKeys.Remove(foreignKeyOperation);
+                        alterOperations.Add(foreignKeyOperation);
+                    }
+                    return true;
+                })
+                .Select(k => createTableOperations[k]);
```
Comment (bricelam) [inline src/EntityFramework.Relational/Migrations/Infrastructure/ModelDiffer.cs]: I'd call `.ToList()` on this. It seems bad to modify the `alterOperations` collection as it lazily evals in the return statement.

Comment (mikary) [inline src/EntityFramework.Relational/Migrations/Infrastructure/ModelDiffer.cs]: The ```TopologicalSort``` call isn't lazy so ```ToList()``` would only really impact the ```Select(k => createTableOperations[k])``` which is fetching from a dictionary. That being said the lazyness is an implementation detail and not necessarily part of the contract so I could see adding it just to be explicit.

**Pull request merged**